scipy/sparse/_base.py

```diff
@@ -1165,8 +1165,13 @@ def sum(self, axis=None, dtype=None, out=None):
                 np.ones((N, 1), dtype=res_dtype)
             )
 
-        if out is not None and out.shape != ret.shape:
-            raise ValueError("dimensions do not match")
+        if out is not None:
+            if isinstance(self, sparray):
+                if out.shape != ret.squeeze().shape:
```

Comment:
This will do the wrong thing for a an `(N, 1)` sparse matrix/array where we sum over the first axis. `out.shape` should be `(1,)` but ret.shape will be `(1, 1)` and thus the squeezed shape would be `()`.

Reply:
I replaced `squeeze` with code that manually takes out the right number of length 1 dimensions.